File: GeoQuiz/app/src/main/java/com/davdo/geoquiz/src/Question.java
package com.davdo.geoquiz.src;

import android.os.Parcel;
import android.os.Parcelable;

public class Question implements Parcelable {

    private int mTextResId;
    private boolean mCorrectAnswer;

    private int mUserAnswer;
    private boolean mUserCheated;

    public Question(int textResId, boolean answer) {
        mTextResId = textResId;
        mCorrectAnswer = answer;
        mUserAnswer = 0;
    }

    protected Question(Parcel in) {
        mTextResId = in.readInt();
        mCorrectAnswer = (in.readByte() != 0);
        mUserAnswer = in.readInt();
        mUserCheated = (in.readByte() != 0);
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(mTextResId);
        dest.writeByte((byte) (mCorrectAnswer ? 1 : 0));
        dest.writeInt(mUserAnswer);
        dest.writeByte((byte) (mUserCheated ? 1 : 0));
    }

    public void resetAnswer() {
        mUserAnswer = 0;
    }

    public void setUserAnswer(boolean inAnswer) {
        mUserAnswer = inAnswer ? 1 : -1;
    }

    public boolean getUserAnswer() {

        return mUserAnswer == 1;
    }

    public boolean hasUserAnswered() {
        return (mUserAnswer != 0);
    }

    public void setUserCheated(boolean cheated) {
        mUserCheated = cheated;
    }

    public boolean hasUserCheated() {
        return mUserCheated;
    }

    //This will return false when the user has not answered.
    public boolean isUserCorrect() {
      return (mCorrectAnswer ? 1 : -1) == mUserAnswer;
    }

    public int getTextResId() {
        return mTextResId;
    }

    public boolean getCorrectAnswer() {
        return mCorrectAnswer;
    }

    public void setTextResId(int textResId) {
        mTextResId = textResId;
    }

    public void setCorrectAnswer(boolean answerTrue) {
        mCorrectAnswer = answerTrue;
    }



    @Override
    public int describeContents() {
        return 0;
    }

    // --- Static Methods ---

    public static final Creator<Question> CREATOR = new Creator<Question>() {
        @Override
        public Question createFromParcel(Parcel in) {
            return new Question(in);
        }

        @Override
        public Question[] newArray(int size) {
            return new Question[size];
        }
    };
}

File: GeoQuizKotlin/app/src/main/java/com/davdo/geoquizkotlin/src/Question.kt
package com.davdo.geoquizkotlin.src

import android.os.Parcel
import android.os.Parcelable

class Question : Parcelable {

    var textResId: Int = 0
    var correctAnswer: Boolean = false
    var userCheated: Boolean = false

    private var mUserAnswer: Int = 0

    constructor (resId: Int, answer: Boolean) {
        textResId = resId
        mUserAnswer = 0
        correctAnswer = answer
    }

    constructor(arrive: Parcel) {

        textResId = arrive.readInt()
        correctAnswer = arrive.readInt() != 0
        userCheated = arrive.readInt() != 0
        mUserAnswer = arrive.readInt()
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {

        dest.writeInt(textResId)
        dest.writeInt((if (correctAnswer) 1 else 0))
        dest.writeInt((if (userCheated) 1 else 0))
        dest.writeInt(mUserAnswer)
    }

    fun resetAnswer()  {
        mUserAnswer = 0
    }

    fun setUserAnswer(inAnswer: Boolean) {
        mUserAnswer = if (inAnswer) 1 else -1
    }

    fun getUserAnswer(): Boolean {
        return mUserAnswer == 1
    }

    fun hasUserAnswered(): Boolean {
        return mUserAnswer != 0
    }

    //This will return false when the user has not answered.
    fun isUserCorrect(): Boolean {
        return (if (correctAnswer) 1 else -1) == mUserAnswer
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<Question> {
        override fun createFromParcel(parcel: Parcel): Question {
            return Question(parcel)
        }

        override fun newArray(size: Int): Array<Question?> {
            return arrayOfNulls(size)
        }
    }
}